python/ffsim/tenpy/hamiltonians/molecular_hamiltonian.py

```diff
@@ -0,0 +1,143 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+from __future__ import annotations
+
+import numpy as np
+from tenpy.models.lattice import Lattice
+from tenpy.models.model import CouplingMPOModel
+from tenpy.networks.site import SpinHalfFermionSite
+
+from ffsim.hamiltonians.molecular_hamiltonian import MolecularHamiltonian
+
+# ignore lowercase variable checks to maintain TeNPy naming conventions
+# ruff: noqa: N806
+
+
+class MolecularHamiltonianMPOModel(CouplingMPOModel):
+    """Molecular Hamiltonian."""
+
+    def init_sites(self, params):
+        cons_N = params.get("cons_N", "N")
+        cons_Sz = params.get("cons_Sz", "Sz")
+        site = SpinHalfFermionSite(cons_N=cons_N, cons_Sz=cons_Sz)
+        return site
+
+    def init_lattice(self, params):
+        L = params.get("L", 1)
+        norb = params.get("norb", 4)
+        site = self.init_sites(params)
+        basis = np.array(([norb, 0.0], [0, 1]))
+        pos = np.array([[i, 0] for i in range(norb)])
+        lat = Lattice(
+            [L, 1],
+            [site] * norb,
+            order="default",
+            bc="open",
+            bc_MPS="finite",
+            basis=basis,
+            positions=pos,
+        )
+        return lat
+
+    def init_terms(self, params):
+        dx0 = np.array([0, 0])
+        norb = params.get("norb", 4)
+        one_body_tensor = params.get("one_body_tensor", np.zeros((norb, norb)))
+        two_body_tensor = params.get(
+            "two_body_tensor", np.zeros((norb, norb, norb, norb))
+        )
+        constant = params.get("constant", 0)
+
+        for p in range(norb):
+            for q in range(norb):
+                h1 = one_body_tensor[q, p]
+                if p == q:
+                    self.add_onsite(h1, p, "Nu")
+                    self.add_onsite(h1, p, "Nd")
+                    self.add_onsite(constant / norb, p, "Id")
+                else:
+                    self.add_coupling(h1, p, "Cdu", q, "Cu", dx0)
+                    self.add_coupling(h1, p, "Cdd", q, "Cd", dx0)
+
+                for r in range(norb):
+                    for s in range(norb):
+                        h2 = two_body_tensor[q, p, s, r]
+                        if p == q == r == s:
+                            self.add_onsite(h2 / 2, p, "Nu")
```

Comment:
Nit: Prefer multiplication when it's exact, since it's usually faster than division. I.e. `0.5 * h2`